src/oscar/apps/catalogue/views.py

```diff
@@ -170,14 +170,35 @@ def get(self, request, *args, **kwargs):
     def get_category(self):
         if 'pk' in self.kwargs:
             # Usual way to reach a category page. We just look at the primary
-            # key in case the slug changed. If it did, get() will redirect
-            # appropriately
-            filters = {'pk': self.kwargs['pk']}
-        else:
-            # For SEO reasons, we allow chopping off bits of the URL. If that
-            # happened, no primary key will be available.
-            filters = {'slug': self.kwargs['category_slug']}
-        return get_object_or_404(Category, **filters)
+            # key, which is easy on the database. If the slug changed, get()
+            # will redirect appropriately.
+            # WARNING: Category.get_absolute_url needs to look up it's parents
+            # to compute the URL. As this is slightly expensive, Oscar's
+            # default implementation caches the method. That's pretty safe
+            # as ProductCategoryView does the lookup by primary key, which
+            # will work even if the cache is stale. But if you override this
+            # logic, consider if that still holds true.
+            return get_object_or_404(Category, pk=self.kwargs['pk'])
+        elif 'category_slug' in self.kwargs:
+            # For SEO and legacy reasons, we allow chopping off the primary
+            # key from the URL. In that case, we have the target category slug
+            # and it's ancestors' slugs concatenated together.
+            # To save on queries, we pick the last slug, look up all matching
+            # categories and only then compare.
+            # Note that currently we enforce uniqueness of slugs, but as that
+            # might feasibly change soon, it makes sense to be forgiving here.
+            concatenated_slugs = self.kwargs['category_slug']
+            slugs = concatenated_slugs.split(Category._slug_separator)
+            try:
+                last_slug = slugs[-1]
+            except IndexError:
+                raise Http404
+            else:
+                for category in Category.objects.filter(slug=last_slug):
+                    if category.full_slug == concatenated_slugs:
+                        return category
```

Comment:
Ah, indeed, I missed the fact that you compare against `full_slug`. Should be alright then :)

I'd be alright with removing the slug-only variant, but I'd check history to see whether there's more clue as to why that was introduced in the first place.